google_symptoms/delphi_google_symptoms/pull.py

```diff
@@ -0,0 +1,113 @@
+    conforms it into a dataset
+
+    The output dataset has:
+
+    - Each row corresponds to (County/State, Date),
+      denoted (FIPS/State abbrev, timestamp).
+    - Each row additionally has columns corresponding to sensors such as
+      "Anosmia" and "Ageusia".
+
+    Note that we retrieve state level data from "2020_US_daily_symptoms_dataset.csv"
+    where there are state level data for 51 states including 'District of Columbia'.
+
+    We retrieve the county level data from "/subregions/state/**daily**.csv"
+    where there is county level data available except District of Columbia.
+    We filter the data such that we only keep rows with valid FIPS.
+
+    PS:  No information for PR
+
+    Parameters
+    ----------
+    base_url: str
+        Base URL for pulling the Google COVID-19 Search Trends symptoms dataset
+    metrics: list of string
+        Symptoms to consider: "Anosmia" and "Ageusia".
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe as described above.
+    """
+    # Constants
+    KEEP_COLUMNS = ["geo_id", "date"]
+    for metric in metrics:
+        KEEP_COLUMNS.append("symptom:" + metric)
+
+    # Read data
+    if level == "state":
+        df = pd.read_csv(base_url.format(sub_url="/", state=""),
+                         parse_dates = ["date"])
+        df["geo_id"] = df["open_covid_region_code"].apply(
+                lambda x: x.split("-")[1].lower()
+        )
+
+    else:
+        dfList = []
+        for state in list(STATE_TO_ABBREV.keys()):
+            sub_url = "/subregions/" + "%20".join(state.split("_")) + "/"
+            dfList.append(pd.read_csv(base_url.format(sub_url=sub_url,
+                                                      state=state+"_"),
+                                      parse_dates = ["date"]))
+        df = pd.concat(dfList)
+        df["geo_id"] = df["open_covid_region_code"].apply(get_geo_id)
+
+    # Delete rows with missing FIPS
+    null_mask = (df["geo_id"].isnull())
+    df = df.loc[~null_mask]
+
+    # Confirm FIPS
+    if level == "county":
+        test_result = []
+        for fips in df["geo_id"].values:
+            test_result.append(len(re.findall(r"\b\d{5}\b",fips)) == 1)
+        assert all(test_result)
+
+    # keep necessary columns only
+    try:
+        df = df[KEEP_COLUMNS]
+    except KeyError:
+        raise ValueError(
+            "Part of necessary columns are missed. The dataset "
+            "schema may have changed.  Please investigate."
+        )
+
+    # Let each FIPS/state has same number of rows
+    geo_list = df["geo_id"].unique()
+    date_list = pd.date_range(start=df["date"].min(),
+                              end=df["date"].max(),
+                              freq='D')
+    index_df = pd.MultiIndex.from_product(
+        [geo_list, date_list], names=['geo_id', 'date']
+    )
+    df = df.groupby(
+            ["geo_id", "date"]
+        ).sum().reindex(
```

Comment:
The raw series are ending up with missing values overwritten by 0. Looks like it's coming from the `sum()` here, which defaults to ignoring `NaN`s. Since this is desired behavior in the smoothed case, so we'll probably need an `if level == "state"` around here.